lib/sectorbuilder/fs.go

```diff
@@ -0,0 +1,115 @@
+package sectorbuilder
+
+import (
+	"github.com/filecoin-project/lotus/chain/types"
+	"golang.org/x/xerrors"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+)
+
+type dataType int
+
+const (
+	dataCache dataType = iota
+	dataStaging
+	dataSealed
+	dataUnsealed
+
+	nDataTypes
+)
+
+var overheadMul = []uint64{ // * sectorSize
+	dataCache:    11, // TODO: check if true for 32G sectors
+	dataStaging:  1,
+	dataSealed:   1,
+	dataUnsealed: 1,
+}
+
+type fs struct {
+	path string
+
+	// in progress actions
+
+	reserved [nDataTypes]uint64
+
+	lk sync.Mutex
+}
+
+func openFs(dir string) *fs {
+	return &fs{
+		path: dir,
+	}
+}
+
+func (f *fs) init() error {
+	for _, dir := range []string{f.path, f.cache(), f.staging(), f.sealed(), f.unsealed()} {
+		if err := os.Mkdir(dir, 0755); err != nil {
+			if os.IsExist(err) {
+				continue
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *fs) cache() string {
+	return filepath.Join(f.path, "cache")
+}
+
+func (f *fs) staging() string {
+	return filepath.Join(f.path, "staging")
+}
+
+func (f *fs) sealed() string {
+	return filepath.Join(f.path, "sealed")
+}
+
+func (f *fs) unsealed() string {
+	return filepath.Join(f.path, "unsealed")
+}
+
+func (f *fs) reservedBytes() int64 {
+	var out int64
+	for _, r := range f.reserved {
+		out += int64(r)
+	}
+	return out
+}
+
+func (f *fs) reserve(typ dataType, size uint64) error {
+	f.lk.Lock()
+	defer f.lk.Unlock()
+
+	var fsstat syscall.Statfs_t
+
+	if err := syscall.Statfs(f.path, &fsstat); err != nil {
```

Comment:
lets stat the subdirectory, on my computer I map the staged and cache directory to my SSD, and the sealed to an nfs mount